spec/services/bulk_sql_delete_spec.rb

```diff
@@ -0,0 +1,51 @@
+require "rails_helper"
+
+describe BulkSqlDelete do
+  let(:sql) do
+    <<-SQL
+      DELETE FROM notifications
+      WHERE notifications.id IN (
+        SELECT notifications.id
+        FROM notifications
+        WHERE created_at < '#{Time.zone.now}'
+        LIMIT 1
+      )
+    SQL
+  end
+  let(:bulk_deleter) { subject }
+  let(:logger) { Rails.logger }
+
+  before { allow(Rails).to receive(:logger).and_return(logger) }
+
+  describe "#delete_in_batches" do
+    it "logs batch deletion" do
+      create_list :notification, 5
+      allow(logger).to receive(:info).exactly(6).times.with(
+        hash_including(:tag, :statement, :duration, :rows_deleted),
+      )
+      bulk_deleter.delete_in_batches(sql)
+      expect(logger).to have_received(:info).exactly(6).times.with(
+        hash_including(:tag, :statement, :duration, :rows_deleted),
+      )
+    end
+
+    it "logs errors that occur" do
+      allow(logger).to receive(:error).with(
+        hash_including(:tag, :statement, :exception_message, :backtrace),
+      )
+      # rubocop:disable RSpec/AnyInstance
+      allow_any_instance_of(ActiveRecord::ConnectionAdapters::PostgreSQLAdapter).to receive(:exec_delete).and_raise("broken")
```

Comment:
I used `allow_any_instance_of` bc I went down a seemingly endless rabbit hole trying to mock out the adapter module and all of its functionality and this seemed like a much easier solution. If anyone wants to offer an alternative solution let me know!

Reply:
You could cheat and mock one of the private methods but I'm not sure it'd be better :(